paddle/fluid/operators/take_along_axis_op.cc

```diff
@@ -0,0 +1,142 @@
+  }
+};
+
+class TakeAlongAxisOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  void Make() override {
+    AddInput("Input", "The input tensor of TakeAlongAxisOp");
+    AddInput("Index", "The index tensor of TakeAlongAxisOp");
+    AddOutput("Result", "The result tensor of TakeAlongAxisOp");
+    AddAttr<int>("Axis",
+                 "The Tensor which contains the axis that we do TakeAlongAxis "
+                 "operation.");
+    AddComment(R"DOC(
+        Take_along_axis Operator.)
+    )DOC");
+  }
+};
+
+class TakeAlongAxisGradOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+  void InferShape(framework::InferShapeContext* ctx) const override {
+    ctx->SetOutputDim(framework::GradVarName("Input"),
+                      ctx->GetInputDim("Input"));
+    ctx->ShareLoD("Input", /*-->*/ framework::GradVarName("Input"));
+  }
+
+ protected:
+  framework::OpKernelType GetExpectedKernelType(
+      const framework::ExecutionContext& ctx) const override {
+    return framework::OpKernelType(OperatorWithKernel::IndicateVarDataType(
+                                       ctx, framework::GradVarName("Result")),
+                                   ctx.device_context());
+  }
+  framework::OpKernelType GetKernelTypeForVar(
+      const std::string& var_name, const framework::Tensor& tensor,
+      const framework::OpKernelType& expected_kernel_type) const override {
+    return framework::OpKernelType(expected_kernel_type.data_type_,
+                                   tensor.place(), tensor.layout());
+  }
+};
+
+template <typename T>
+class TakeAlongAxisGradOpMaker : public framework::SingleGradOpMaker<T> {
+ public:
+  using framework::SingleGradOpMaker<T>::SingleGradOpMaker;
+
+ protected:
+  void Apply(GradOpPtr<T> op) const override {
+    op->SetType("take_along_axis_grad");
+    op->SetInput("Index", this->Input("Index"));
+    op->SetInput("Input", this->Input("Input"));
+
+    op->SetInput(framework::GradVarName("Result"), this->OutputGrad("Result"));
+    op->SetOutput(framework::GradVarName("Input"), this->InputGrad("Input"));
+    op->SetAttrMap(this->Attrs());
+  }
+};
+
+}  // namespace operators
+}  // namespace paddle
+
+namespace ops = paddle::operators;
+REGISTER_OPERATOR(take_along_axis, ops::TakeAlongAxisOp,
+                  ops::TakeAlongAxisOpMaker,
+                  ops::TakeAlongAxisGradOpMaker<paddle::framework::OpDesc>,
+                  ops::TakeAlongAxisGradOpMaker<paddle::imperative::OpBase>);
+
+REGISTER_OPERATOR(take_along_axis_grad, ops::TakeAlongAxisGradOp);
+
+REGISTER_OP_CPU_KERNEL(take_along_axis, ops::TakeAlongAxisOpKernel<float>,
+                       ops::TakeAlongAxisOpKernel<double>,
+                       ops::TakeAlongAxisOpKernel<int>,
+                       ops::TakeAlongAxisOpKernel<uint8_t>,
+                       ops::TakeAlongAxisOpKernel<int64_t>);
```

Comment:
Is it possible to register for complex64 and complex128?

Reply:
I don't think so, since the python API level won't support complex type.

Reply:
Why complex type is not supported in python?